### PR TITLE
OWLS-68695 Move load-balance setup to separate helm chart

### PR DESCRIPTION
**Pull request by lilyhe123**
1. Move load balancer related manifest from the domain chart to a new chart loadBalancer-per-domain. 
2. Provide  a shell script to install IngressController for Traefik and Voyager with helm chart.
2. Modify the integration tests accordingly.
3. Use an env variable LB_TYPE to control load balancer type in the integration tests. We support four types: NONE, APACHE, TRAEFIK and VOYAGER.

One behavior change is that in integration tests before this change we create a Traefik IngressController for each domain. With this change only one Traefik IngressController created for all domains and use different host names to route to different domains.

jenkins jobs passed:
LB_TYPE=TRAEFIK: http://wls-jenkins.us.oracle.com/view/weblogic-operator/job/weblogic-kubernetes-operator-javatest/190/.
LB_TYPE=VOYAGER: http://wls-jenkins.us.oracle.com/view/weblogic-operator/job/weblogic-kubernetes-operator-javatest/191/.
LB_TYPE=NONE: http://wls-jenkins.us.oracle.com/view/weblogic-operator/job/weblogic-kubernetes-operator-javatest/194/.

### CODE REVIEW

**Comment by vanajamukkara** [comment]
Do you have a wercker run that passed with these changes?

**Comment by moreaut** [comment]
What is the relationship between this pull request and Simon's sample apache load balancer pull request - https://github.com/oracle/weblogic-kubernetes-operator/pull/426 ?

Also, the helm chart in in this change isn't following the conventions that the operator helm chart uses (e.g. for validating inputs) - the domain .tpl files that this change removed did follow those conventions.

Overall, I'm a bit confused about the requirements (e.g. Simon's sample chart v.s. this one seems to
be part of the operator), and whether this is just a temporary chart or if it's in the final form for 2.0.

**Comment by lilyhe123** [comment]
> Do you have a wercker run that passed with these changes?

No. Can you inform me how to trigger a wrecker run with a branch?

**Comment by russgold** [comment]
It is automatically triggered when you push to the repo. Your latest run <https://app.wercker.com/Oracle/weblogic-kubernetes-operator/runs/build/5bb06e045e7eac00074e67bf?step=5bb06e23f07af0000750a21c> shows unit test failures, because the unit tests in question are no longer relevant, and should be deleted.

**Comment by lilyhe123** [comment]
> What is the relationship between this pull request and Simon's sample apache load balancer pull request - #426 ?

This change is to focus on moving LB-related manifest from domain chart to a separate loadBalacner chart. And when Simon's apache chart is ready, we can move apache-related manifest from the loadBalancer chart to that apache chart.

> Also, the helm chart in this change isn't following the conventions that the operator helm chart uses (e.g. for validating values.yaml) - the domain .tpl files that this change removed did follow those conventions.

Can you explain more about the chart conventions? or any doc/wiki about it?

> Overall, I'm a bit confused about the requirements (e.g. Simon's chart is a sample v.s. this one seems to be part of the operator), and whether this is just a temporary chart or if it's in the final form for 2.0.

I'm working on move the  loadBalacner chart to sample.

**Comment by lilyhe123** [comment]
> It is automatically triggered when you push to the repo. Your latest run <https://app.wercker.com/Oracle/weblogic-kubernetes-operator/runs/build/5bb06e045e7eac00074e67bf?step=5bb06e23f07af0000750a21c> shows unit test failures, because the unit tests in question are no longer relevant, and should be deleted.
> […](#)

Thanks, Russ. Why the jenkins run didn't catch the UT failures? I think this need to be fixed.

**Comment by russgold** [comment]
Do you have a link to the Jenkins run?

**Comment by lilyhe123** [comment]
> Do you have a wercker run that passed with these changes?

jenkins jobs passed:
LB_TYPE=TRAEFIK: http://wls-jenkins.us.oracle.com/view/weblogic-operator/job/weblogic-kubernetes-operator-javatest/177/.
LB_TYPE=VOYAGER: http://wls-jenkins.us.oracle.com/view/weblogic-operator/job/weblogic-kubernetes-operator-javatest/178/.
LB_TYPE=APACHE: http://wls-jenkins.us.oracle.com/view/weblogic-operator/job/weblogic-kubernetes-operator-javatest/183/.
LB_TYPE=NONE: http://wls-jenkins.us.oracle.com/view/weblogic-operator/job/weblogic-kubernetes-operator-javatest/180/

But somehow the server is down today.

**Comment by lilyhe123** [comment]
I remove all apache manifest from this change and disable apache test for now. I assume Simon's PR #426 will enable the apache test and make it pass.

**Comment by lilyhe123** [comment]
@moreaut The ingress-per-domain is very simple. Its only has two template yaml, one for traefik and one for voyager.

I can see some benefits to use your pattern, input validation etc, but the downside is that it will increase the complexity of the chart. It make it difficult to understand and maintain without deep understanding of the Helm template language. 

I'd prefer to leave the simple ingress-per-domain as it is.

**Comment by moreaut** [comment]
I agree with Russell - at the very least, you should temporarily add @Ignore to the tests that are failing instead of disabling all of the tests.

File: integration-tests/src/test/resources/loadBalancer/setup.sh [hunk 1]
```diff
@@ -0,0 +1,138 @@
+#!/bin/bash
+
+#  This script is to create or delete Ingress controllers.
+#  Copyright 2017, 2018, Oracle Corporation and/or affiliates.  All rights reserved.
```
Comment (doxiao) [inline integration-tests/src/test/resources/loadBalancer/setup.sh]: This looks likes a new file. If so, remove "2017" from the Copyright.

File: integration-tests/src/test/resources/loadBalancer/traefik-values.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+serviceType: NodePort
```
Comment (doxiao) [inline integration-tests/src/test/resources/loadBalancer/traefik-values.yaml]: This looks likes a new file. If so,add a Copyright?

File: kubernetes/samples/charts/ingress-per-domain/values.yaml [hunk 3]
```diff
@@ -0,0 +1,24 @@
+# Default values for ingress-per-domain.
```
Comment (doxiao) [inline kubernetes/samples/charts/ingress-per-domain/values.yaml]: add a Copyright?

**Comment by doxiao** [comment]
It would be helpful if we could have a README to show the steps that the user needs to take to start up a load balancer of each type. We have been asked to provide README for other new sample scripts.

**Pull request closed without merging**